File: apps/java/io/dustin/apps/common/advice/GlobalExceptionHandler.java
package io.dustin.apps.common.advice;

import io.dustin.apps.common.exception.BadRequestParameterException;
import io.dustin.apps.common.exception.DataNotFoundException;
import io.dustin.apps.common.utils.SelectDate;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import static java.time.LocalDateTime.now;

@Slf4j
@RestControllerAdvice
public class GlobalExceptionHandler {

    @ResponseStatus(HttpStatus.NOT_FOUND)
    @ExceptionHandler(DataNotFoundException.class)
    protected ApiError handleDataNotFoundException(DataNotFoundException e) {
        log.error(e.getMessage());
        return ApiError.builder()
                .status(HttpStatus.NOT_FOUND.value())
                .message(e.getMessage())
                .timestamp(SelectDate.TYPE_FOUR.transform(now()))
                .build();
    }

    @ResponseStatus(HttpStatus.BAD_REQUEST)
    @ExceptionHandler(BadRequestParameterException.class)
    protected ApiError handleBadRequestParameterException(BadRequestParameterException e) {
        log.error(e.getMessage());
        return ApiError.builder()
                .status(HttpStatus.BAD_REQUEST.value())
                .message(e.getMessage())
                .timestamp(SelectDate.TYPE_FOUR.transform(now()))
                .build();
    }

    @Builder
    @Getter
    @AllArgsConstructor
    public static class ApiError {
        private int status;
        private String message;
        private String timestamp;
    }

}

File: apps/src/main/kotlin/io/dustin/apps/common/advice/GlobalExceptionHandler.kt
package io.dustin.apps.common.advice

import SelectDate
import io.dustin.apps.common.exception.BadRequestParameterException
import io.dustin.apps.common.exception.DataNotFoundException
import io.dustin.apps.common.exception.DuplicatedException
import io.dustin.apps.common.exception.NotFoundEntityException
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.server.ResponseStatusException
import java.time.LocalDateTime.now
import javax.naming.AuthenticationException


@RestControllerAdvice
class GlobalExceptionHandler {

    @ResponseStatus(HttpStatus.CONFLICT)
    @ExceptionHandler(DuplicatedException::class)
    fun handleDuplicatedExceptionException(ex: DuplicatedException): ApiError {
        return ApiError(
            code = HttpStatus.CONFLICT.value(),
            message = ex.message!!,
            timestamp = SelectDate.TYPE_FOUR.transform(now()),
        )
    }

    @ResponseStatus(HttpStatus.NOT_FOUND)
    @ExceptionHandler(NotFoundEntityException::class)
    fun handleNotFoundEntityException(ex: NotFoundEntityException): ApiError {
        return ApiError(
            code = HttpStatus.NOT_FOUND.value(),
            message = ex.message!!,
            timestamp = SelectDate.TYPE_FOUR.transform(now()),
        )
    }

    @ResponseStatus(HttpStatus.NOT_FOUND)
    @ExceptionHandler(DataNotFoundException::class)
    fun handleDataNotFoundException(e: DataNotFoundException): ApiError {
        return ApiError(
            code = HttpStatus.NOT_FOUND.value(),
            message = e.message,
            timestamp = SelectDate.TYPE_FOUR.transform(now())
        )
    }

    @ResponseStatus(HttpStatus.BAD_REQUEST)
    @ExceptionHandler(BadRequestParameterException::class)
    fun handleBadRequestParameterException(e: BadRequestParameterException): ApiError {
        return ApiError(
            code = HttpStatus.BAD_REQUEST.value(),
            message = e.message,
            timestamp = SelectDate.TYPE_FOUR.transform(now())
        )
    }

    @ResponseStatus(HttpStatus.BAD_REQUEST)
    @ExceptionHandler(ResponseStatusException::class)
    fun handleBadRequestParameterException(e: ResponseStatusException): ApiError {
        return ApiError(
            code = HttpStatus.BAD_REQUEST.value(),
            message = e.message,
            timestamp = SelectDate.TYPE_FOUR.transform(now())
        )
    }

    @ResponseStatus(HttpStatus.UNAUTHORIZED)
    @ExceptionHandler(AuthenticationException::class)
    fun handleAuthenticationException(e: ResponseStatusException): ApiError {
        return ApiError(
            code = HttpStatus.BAD_REQUEST.value(),
            message = e.message,
            timestamp = SelectDate.TYPE_FOUR.transform(now())
        )
    }

    data class ApiError(
        val code: Int,
        val message: String?,
        val timestamp: String?
    )
}


